File: COERbuoy/examples/custom_controller/controller.m
clear -all
pkg load instrument-control
clc
 

function [ctrl]=control(status)
%status.time          - a sequence of 100 values, the last element contains the most recent value
%status.wave          - a sequence of 100 values with the wave elevation corresponding to the time series*
%status.wave_forecast - a sequence of 100 values with the wave forecast*
%status.stroke_pos    - a sequence of 100 values with the stroke position; correspoding to the time series
%status.stroke_speed  - a sequence of 100 values with the stroke speed; corresponding to the time series
%status.angular_pos   - a sequence of 100 values with the angular position; corresponding to the time series
%status.angular_speed - a sequence of 100 values with the angular speed; corresponding to the time series
%status.force         - a sequence of 100 values with the force measured by the force sensor; corresponding to the time series
%* not available in the COERbuoy1 benchmark

gamma=100000; %[Ns/m] constant damping value
ctrl.F_pto = -gamma*status.stroke_speed(end); #set PTO force
ctrl.F_brake = 0; #set brake force

endfunction



%%-- do not change --%%

pause(4);
global BUFSIZE=sizeof(struct('time',zeros(100,1),'wave',zeros(100,1),'wave_forecast',zeros(100,1),'stroke_pos',zeros(100,1),'stroke_speed',zeros(100,1),'angular_pos',zeros(100,1),'angular_speed',zeros(100,1),'force',zeros(100,1),'test',zeros(100,1)));
global endian='B';
global s = tcpip('localhost', 5050, 'timeout', 100);
[type, mSize, endian] = computer
function [status]=receivemsg()
    global s;
    global endian;
    global BUFSIZE;
    array=tcp_read(s,BUFSIZE,10000);
    array=typecast(array,'double');
    if endian == 'L';
      array=swapbytes(array);
    endif
    array=array;
      
    i=1;
    status.time=array(1+(i-1)*100:i*100);
    i=i+1;
    status.wave=array(1+(i-1)*100:i*100);
    i=i+1;
    status.wave_forecast=array(1+(i-1)*100:i*100);
    i=i+1;
    status.stroke_pos=array(1+(i-1)*100:i*100);
    i=i+1;
    status.stroke_speed=array(1+(i-1)*100:i*100);
    i=i+1;
    status.angular_pos=array(1+(i-1)*100:i*100);
    i=i+1;
    status.angular_speed=array(1+(i-1)*100:i*100);
    i=i+1;
    status.force=array(1+(i-1)*100:i*100);
    i=i+1;
    status.test=array(1+(i-1)*100:i*100);
endfunction

%send control message
function sendmsg(time,force,brake,test)
    global endian;
    global s;
    array=double([time(1:9),force(1:9),brake(1:9),test(1:9)]);
    if endian == 'L'
      array=swapbytes(array);
    endif
    array=typecast(array,'uint8');
    tcp_write(s,uint8(array));
endfunction

%main
while 1
  #Constant damping control
  status=receivemsg();
  [ctrl]=control(status);
  sendmsg(linspace(status.time(end),status.time(end)+1,9),ones(1,9)*ctrl.F_pto,ones(1,9)*ctrl.F_brake,zeros(1,9));
endwhile

File: examples/visual_controller/controller.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Simple linear damping
@author: Simon H. Thomas, COER laboratory, Maynooth University
"""
import numpy as np;
from scipy.interpolate import interp1d;
import COERbuoy.connection as connection;

conn_model=connection.connection();#Initialize connection
conn_model.openC();#Use client mode
    
while msg:=conn_model.get_control():

  ##Read incoming message
  time  =msg["time"]          #1x100 array with time series
  wave  =msg["wave"];         #1x100 array with wave data (related to time series)*
  wave_f=msg["wave_forecast"];#1x100 array with wave forecast*
  x     =msg["stroke_pos"];   #1x100 array with stroke position (related to time series)
  dx    =msg["stroke_speed"]; #1x100 array with stroke speed (related to time series)
  alpha =msg["angular_pos"];  #1x100 array with pitch angle (related to time series)
  dalpha=msg["angular_speed"];#1x100 array with pitch angular speed (related to time series)
  force =msg["force"];        #1x100 array with force sensor data (related to time series)
  #* not available during COERbuoy1 benchmark
  now=time[-1]; #The last element contains the most recent data (except the wave forecast)
  
  ##This part contains the controller's logic
  ##TODO: replace logic with own controller idea
  
  #(here we use a simple linear damping)
  #Calculate PTO force
  gamma=100000;#kNs/m
  F_pto=-gamma*dx[-1];
  #In this example we don't use the WECs brake
  brake=0;#kNs/m
  
  ##Write control message 
  answer={
          "time":np.linspace(now,now+1,9),#1x9 array with time steps in the future
          "pto":np.array([F_pto]*9),      #1x9 array with PTO force
          "brake":np.zeros(9),            #1x9 array with brake force
          "test":np.zeros(9),             #1x9 array; not used
          }
  #Send control message to model
  conn_model.set_control(answer["time"],answer["pto"],answer["brake"],answer["test"]);

conn_model.close();
